coalib/core/PersistentHash.py

```diff
@@ -2,18 +2,50 @@
 import pickle
 
 
-def persistent_hash(obj):
+def order(obj):
+    """
+    Recursively preserves the order of any sets or dicts
+    found in obj.
     """
-    Calculates a persistent hash of the given object.
+    if isinstance(obj, set):
+        return sorted(obj)
+    elif isinstance(obj, dict):
+        for k, v in obj.items():
+            if isinstance(v, set) or isinstance(v, dict):
+                obj[k] = order(v)
+        return sorted(obj.items())
+    else:
+        return obj
 
-    This hash method uses pickle (protocol 4) to serialize the given object and
-    hash the byte-stream. The hashing algorithm used is SHA-1.
 
-    :param obj:
-        The object to calculate the persistent hash for.
-    :return:
-        The persistent hash.
+class PreserveOrder:
+    """
+    Preserves order of task-objects containing sets and dictionaries.
     """
+
+    def __init__(self, arg_tuple, arg_dict):
+        self.arg_tuple = arg_tuple
+        self.arg_dict = arg_dict
+        self.list = []
+        self.dict = {}
+
+    def order_dict(self):
+        for key, value in self.arg_dict.items():
+            self.dict[key] = order(value)
+        return sorted(self.dict.items())
+
+    def order_tuple(self):
+        for o in self.arg_tuple:
+            self.list.append(order(o))
+        return tuple(self.list)
+
+
+def persistent_hash(obj):
+    if isinstance(obj, tuple):
```

Comment:
why checking if it's a tuple before checking anything else? What if I directly pass it a dict? I would like to see it working for them too^^

Reply:
Hm just thinking, may this work?
```python
obj = order(obj)

# do pickle things
```